package-name-map/src/package-name-map.ts

```diff
@@ -0,0 +1,278 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/**
+ * A Scope and also the top-level interface for package name mappings.
+ */
+export interface Scope {
+  path_prefix?: string;
+  packages?: {[name: string]: Package};
+  scopes?: {[path: string]: Scope};
+}
+
+export interface Package {
+  main: string;
+  path?: string;
+}
+
+/**
+ * A path/Scope pair for use in the internal algorithm.
+ */
+interface ScopeEntry {
+  /**
+   * The absolute prefix URL to be applied to any path within the scope. This
+   * includes this Scope's any ancestor's paths and path_prefixes.
+   */
+  prefixURL: string;
+
+  scope: Scope;
+}
+
+/**
+ * The result of the FindPackageEntry operation.
+ */
+interface FindPackageResult {
+  scopeContext: ScopeEntry[];
+  packageName?: string;
+  package?: Package;
+}
+
+/**
+ * An object that can resolve specifiers using its package name map
+ * configuration object.
+ *
+ * The schema of the configuration object is defined in the proposal.
+ *
+ * Currently this class can only have one mapping, but a future change could
+ * support multiple mappings, added over time, as discussed in
+ * https://github.com/domenic/package-name-maps/issues/14
+ *
+ * This class doesn't yet implement any validation of the mapping. We need to
+ * decide whether validation errors should be early or late errors. Either way,
+ * it's still a TODO to implement a validate() method.
+ */
+export class PackageNameMap {
+  private _baseURI: string;
+  private _map: Scope;
+
+  constructor(map: Scope, baseURI: string) {
+    this._map = map;
+    this._baseURI = new URL('.', baseURI).href;
+  }
+
+  resolve(specifier: string, referrerURL: string): string {
+    // 1. If specifier parses as absolute URL, return the specifier.
+    if (isURL(specifier)) {
+      return specifier;
+    }
+
+    // 2. If specifier starts with `./`, `../`,  or `/`, return the specifier
+    //    resolved from the referrerURL.
+    if (/^\.{0,2}\//.test(specifier)) {
+      return new URL(specifier, referrerURL).href;
+    }
+
+    // 3. Find the applicable scope based on referrerURL.
+    const initialScopeContext = getScopeContext(
+      this._map,
+      referrerURL,
+      this._baseURI
+    );
+
+    // 4. Find the package entry.
+    const {scopeContext, packageName, package: pkg} = findPackage(
+      initialScopeContext,
+      specifier
+    );
+
+    // 5. If no package entry is found, throw an error.
+    if (pkg === undefined) {
+      throw new Error(
+        `Unable to resolve specifier ${specifier} from referrer ${referrerURL}`
+      );
+    }
+
+    // 6. If the specifier is to a sub-module, but the package doesn't have a
+    //    path, throw an error.
+    if (specifier !== packageName && pkg.path === undefined) {
```

Comment:
I understand the sugars are still being fully decided, but if the default path is assumed to be the package name when setting a main, it seems a little inconsistent to then require an explicit path in the subpath case. Perhaps we should do both or neither instead of just doing the main assumption here?